dartagnan/src/main/java/com/dat3m/dartagnan/program/ScopedThread/PTXThread.java

```diff
@@ -0,0 +1,44 @@
+package com.dat3m.dartagnan.program.ScopedThread;
+
+import com.dat3m.dartagnan.program.Thread;
+import com.dat3m.dartagnan.program.event.Tag;
+import com.dat3m.dartagnan.program.event.core.Event;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+
+public class PTXThread extends Thread implements ScopedThread {
```

Comment:
This class seems unnecessary to me. For what I see, you only need to know to which gpu and cta a thread belongs to (AFAIK there is a unique system). This can be replaced by adding final fields to the Thread class, setting it to null when there are no scopes, and implementing two methods `getGpu` and `getCta`.

Reply:
I was thinking to keep the scope feature out of thread for future scoped thread from other architectures. For example, in vulkan the scope is {`subgroup`/`workgroup`/`queuefamily`/`device`}.

Reply:
Ok. that makes sense.
But some small improvements:
- you can have `private final HashMap<String, Integer> scopeIds = new HashMap<>();` ... no need to initialize the map when the constructor is called.
- `getScopes` and `getScopeIds` will probably have the same implementation for any class implementing `ScopedThread`, thus you can simply add this implementation as the default on in the interface

Reply:
If all implementations (`PTXThread` and `VulkanThread`) will have the same implementation of those methods, there is not point in making `ScopedThread` an interface. What about the following?
```
public class ScopedThread extends Thread {

    protected final HashMap<String, Integer> scopeIds = new HashMap<>();
    protected final ArrayList<String> scopes = new ArrayList<>();

    public ScopedThread(String name, int id, Event entry) {
        super(name, id, entry);
    }

    public ArrayList<String> getScopes() {
        return scopes;
    }

    public int getScopeIds(String scope) {
        if (!this.scopeIds.containsKey(scope)) {
            return -1;
        }
        return this.scopeIds.get(scope);
    }
}
```
and
```
public class PTXThread extends ScopedThread {

    public PTXThread(String name, int id, Event entry, int GpuId, int CtaId) {
        super(name, id, entry);
        scopes.addAll(Arrays.asList(Tag.PTX.SYS, Tag.PTX.GPU, Tag.PTX.CTA));
        this.scopeIds.put(Tag.PTX.SYS, 0);
        this.scopeIds.put(Tag.PTX.GPU, GpuId);
        this.scopeIds.put(Tag.PTX.CTA, CtaId);
    }
}
```

Reply:
`this.scopeIds.put(Tag.PTX.SYS, 0);` is useless ... every thread will always have the same id for `SYS` so there is no need to track it

Reply:
Accepted